File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AvoidDateTimeNowForBenchmarking.vb
Imports System

Public Class Program
    Private Sub Benchmark()
        Dim start = Date.Now
        MethodToBeBenchmarked()
        ' something
        Console.WriteLine($"{(Date.Now - start).TotalMilliseconds} ms") ' Noncompliant {{Avoid using "DateTime.Now" for benchmarking or timing operations}}
        '                     ^^^^^^^^^^^^^^^^
    End Sub

    Private Const MinRefreshInterval As Integer = 100

    Private Sub Timing()
        Dim lastRefresh = Date.Now
        If (Date.Now - lastRefresh).TotalMilliseconds > MinRefreshInterval Then ' Noncompliant
        '   ^^^^^^^^^^^^^^^^^^^^^^
            lastRefresh = Date.Now
            ' Refresh
        End If
    End Sub

    Private timeField As TimeSpan

    Public Property Time As TimeSpan
        Get
            Return timeField
        End Get
        Set(ByVal value As TimeSpan)
            Dim start = Date.Now
            MethodToBeBenchmarked()
            timeField = Date.Now - start ' Noncompliant
        End Set
    End Property

    Private Sub SwitchExpression()
        Dim a = 1
        Select Case a
            Case 1
                Dim start = Date.Now
                MethodToBeBenchmarked()
                timeField = Date.Now - start - New TimeSpan(1) ' Noncompliant
            Case 2
        End Select
    End Sub

    Private Function MethodToBeBenchmarked() As Boolean
        Return True
    End Function
End Class

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/AvoidDateTimeNowForBenchmarking.cs
using System;

public class Program
{
    void Benchmark()
    {
        var start = DateTime.Now;
        // Some method
        Console.WriteLine($"{(DateTime.Now - start).TotalMilliseconds} ms"); // Noncompliant {{Avoid using "DateTime.Now" for benchmarking or timespan calculation operations.}}
        //                    ^^^^^^^^^^^^^^^^^^^^

        start = DateTime.Now;
        // Some method
        Console.WriteLine($"{DateTime.Now.Subtract(start).TotalMilliseconds} ms"); // Noncompliant
        //                   ^^^^^^^^^^^^^^^^^^^^^
    }

    private const int MinRefreshInterval = 100;

    void Timing()
    {
        var lastRefresh = DateTime.Now;
        if ((DateTime.Now - lastRefresh).TotalMilliseconds > MinRefreshInterval) // Noncompliant
        {
            lastRefresh = DateTime.Now;
            // Refresh
        }
    }

    void Combinations(TimeSpan timeSpan, DateTime dateTime)
    {
        _ = (DateTime.Now - dateTime).Milliseconds; // Noncompliant
        _ = DateTime.Now.Subtract(dateTime).Milliseconds; // Noncompliant

        _ = (DateTime.Now - TimeSpan.FromSeconds(1)).Millisecond; // Compliant
        _ = DateTime.Now.Subtract(TimeSpan.FromDays(1)).Millisecond; // Compliant

        _ = (DateTime.Now - timeSpan).Millisecond; // Compliant
        _ = DateTime.Now.Subtract(timeSpan).Millisecond; // Compliant

        _ = (DateTime.UtcNow - dateTime).Milliseconds; // Compliant
        _ = DateTime.UtcNow.Subtract(dateTime).Milliseconds; // Compliant

        _ = (new DateTime(1) - dateTime).Milliseconds; // Compliant
        _ = new DateTime(1).Subtract(dateTime).Milliseconds; // Compliant

        void LocalMethod()
        {
            var sec = DateTime.Now;
            // something
            Console.WriteLine($"{(DateTime.Now - sec).TotalMilliseconds} ms"); // Noncompliant
        }
    }

    private TimeSpan time;

    public TimeSpan Time
    {
        get => time;
        set
        {
            var start = DateTime.Now;
            time = DateTime.Now - start; // Noncompliant
        }
    }

    void SwitchExpression(DateTime start)
    {
        var a = 1;
        switch (a)
        {
            case 1:
                time = DateTime.Now - start - new TimeSpan(1); // Noncompliant
                break;
        }
    }

    void NonInLineDateTimeNow()
    {
        var start = DateTime.Now;
        // Some method
        var end = DateTime.Now;
        var elapsedTime = end - start; // FN
    }

    void EdgeCases(DateTime dateTime, TimeSpan timeSpan)
    {
        (true ? DateTime.Now : new DateTime(1)).Subtract(dateTime); // FN
        (true ? DateTime.Now : new DateTime(1)).Subtract(timeSpan); // Compliant

        DateTime.Now.AddDays(1).Subtract(dateTime); // FN
        DateTime.Now.Subtract(); // Error [CS1501]
    }
}

public class FakeDateTimeSubtract
{
    void MyMethod(System.DateTime dateTime)
    {
        _ = DateTime.Now.Subtract(dateTime).Milliseconds; // Compliant
    }

    public static class DateTime
    {
        public static System.DateTime Now { get; }
    }
}
